lib/bigdecimal/math.rb

```diff
@@ -132,6 +133,54 @@ def cos(x, prec)
     y
   end
 
+  # call-seq:
+  #   tan(decimal, numeric) -> BigDecimal
+  #
+  # Computes the tangent of +decimal+ to the specified number of digits of
+  # precision, +numeric+.
+  #
+  # If +decimal+ is Infinity or NaN, returns NaN.
+  #
+  #   BigMath.tan(BigDecimal("0.0"), 4).to_s
+  #   #=> "0.0"
+  #
+  #   BigMath.tan(BigMath.PI(4) / 4, 4).to_s
+  #   #=> "0.999999999999999999999955588155008544487055622030633191403625599381672572e0"
+  #
+  def tan(x, prec)
+    sine = sin(x, prec)
+
+    n    = prec + BigDecimal.double_fig
+    one  = BigDecimal("1")
+    two  = BigDecimal("2")
+    x = -x if x < 0
+    if x > (twopi = two * BigMath.PI(prec))
+      if x > 30
+        x %= twopi
+      else
+        x -= twopi while x > twopi
+      end
+    end
+    x1 = one
+    x2 = x.mult(x,n)
+    sign = 1
+    cosine = one
+    d = cosine
+    i = BigDecimal("0")
+    z = one
+    while d.nonzero? && ((m = n - (cosine.exponent - d.exponent).abs) > 0)
+      m = BigDecimal.double_fig if m < BigDecimal.double_fig
+      sign = -sign
+      x1  = x2.mult(x1,n)
+      i  += two
+      z  *= (i-one) * i
+      d   = sign * x1.div(z,m)
+      cosine += d
+    end
+
+    sine / cosine
+  end
```

Comment:
Could you please mention the origin such as the paper reference of this algorithm in the comment?
And, I want to know how this algorithm is better than just doing `sine/(1 - sine**2)`.  Could you explain?

Reply:
@mrkn I don't believe <tt>tan<var>θ</var></tt> is defined as <tt>sin<var>θ</var> / (1 - sin<sup>2</sup><var>θ</var>)</tt>, but should be <tt>sin<var>θ</var> / cos<var>θ</var></tt> and <tt>cos<var>θ</var></tt> is <tt>√(1 - sin<sup>2</sup><var>θ</var>)</tt>.

Reply:
@nobu The current code was updated against my comment above. Please compare the commit timestamp and the comment timestamp.